File: scripts/scenarios/scenarios/32_identity_fatigue.py
"""Scenario 32: "Identity Fatigue" — PLC Hourly Rate Limits

Test PLC hourly operation rate limits.
The PLC auditor enforces three tiers (configurable via env vars):
  - Hourly: PLC_HOURLY_LIMIT (default 5 in test scenarios)
  - Daily:  PLC_DAILY_LIMIT  (default 15 in test scenarios)
  - Weekly: PLC_WEEKLY_LIMIT (default 50 in test scenarios)

User rosa performs PLC_HOURLY_LIMIT handle rotations.
Assert that the next operation fails with a 400 error from the PLC directory
containing "Too many operations within last hour".

Services: PDS, PLC
"""

from __future__ import annotations

import os
import sys
import time
import requests as reqs
from pathlib import Path

_project_root = str(Path(__file__).resolve().parent.parent.parent.parent)
if _project_root not in sys.path:
    sys.path.insert(0, _project_root)

from scripts.lib.atproto import XrpcClient, get_character, PDS1, ScenarioResult, timed_call


def run() -> ScenarioResult:
    result = ScenarioResult("Identity Fatigue")
    result.start()

    client = XrpcClient(PDS1)
    rosa = get_character("rosa")

    timed_call(result, "PDS health check", lambda: client.wait_for_healthy(timeout=30))
    
    session = timed_call(
        result, "Create account: rosa",
        lambda: client.accounts.create_account(rosa.handle, rosa.email, rosa.password),
        detail_fn=lambda s: f"did={s['did']}",
    )
    if not session:
        result.finish()
        return result
    rosa.did = session["did"]
    rosa.access_jwt = session["accessJwt"]

    # Read the configured hourly limit (set by setup_local_network.sh)
    hourly_limit = int(os.environ.get("PLC_HOURLY_LIMIT", "5"))
    
    # Account creation already consumed one PLC operation, so we can do
    # (hourly_limit - 1) more rotations before hitting the limit.
    rotations = hourly_limit - 1
    print(f"Starting {rotations} handle rotations (account creation already used 1 of {hourly_limit} hourly ops)")
    
    success_count = 0
    try:
        for i in range(rotations):
            new_handle = f"rev-{i}-{rosa.handle}"
            
            # 1. Request signature
            token_resp = client.raw.xrpc_post("com.atproto.identity.requestPlcOperationSignature", {}, token=rosa.access_jwt)
            token = token_resp.get("token")
            
            # 2. Sign operation
            sign_resp = client.raw.xrpc_post("com.atproto.identity.signPlcOperation",
                                            {"token": token, "alsoKnownAs": [f"at://{new_handle}"]},
                                            token=rosa.access_jwt)
            op = sign_resp.get("operation", {})
            op.pop("did", None)
            
            # 3. Submit to PLC
            plc_url = f"http://localhost:2582/{rosa.did}"
            plc_submit = reqs.post(plc_url, json=op, timeout=5)
            
            if plc_submit.status_code == 200:
                success_count += 1
            else:
                result.step_failed("Exhaust Quota", f"Failed at iteration {i}: {plc_submit.status_code} {plc_submit.text}")
                break
            
            if (i + 1) % 5 == 0:
                print(f"  ... completed {i+1} rotations")
    except Exception as exc:
        result.step_failed("Exhaust Quota", str(exc))

    result.step_passed("Quota Exhaustion", f"Successfully performed {success_count} rotations")

    # The next rotation should fail (hourly limit reached)
    if success_count == rotations:
        print(f"Attempting rotation #{rotations + 1} (expect failure — hourly limit reached)")
        new_handle = f"final-{rosa.handle}"
        token_resp = client.raw.xrpc_post("com.atproto.identity.requestPlcOperationSignature", {}, token=rosa.access_jwt)
        token = token_resp.get("token")
        sign_resp = client.raw.xrpc_post("com.atproto.identity.signPlcOperation",
                                        {"token": token, "alsoKnownAs": [f"at://{new_handle}"]},
                                        token=rosa.access_jwt)
        op = sign_resp.get("operation", {})
        op.pop("did", None)
        
        plc_submit = reqs.post(f"http://localhost:2582/{rosa.did}", json=op, timeout=5)
        if plc_submit.status_code == 400 and "Too many operations" in plc_submit.text:
             result.step_passed("Verify Hourly Limit", f"Rejected operation after {hourly_limit} total ops as expected (400)")
        else:
             result.step_failed("Verify Hourly Limit", f"Expected 400 rejection, but got {plc_submit.status_code}: {plc_submit.text}")

    result.finish()
    return result


if __name__ == "__main__":
    res = run()
    res.print_summary()
    sys.exit(res.exit_code)

File: scripts/scenarios/scenarios/32_identity_fatigue.ts
/**
 * @module scenarios/32_identity_fatigue
 *
 * Scenario: Exhausts PLC operation quota and rejects further identity operations.
 *
 * Behavior:
 * - Executes the 32 identity fatigue scenario.
 * - Validates core operations.
 *
 * Expectations:
 * - Scenario completes successfully without errors.
 */

import { getCharacter, PDS1, SERVICE_URLS } from "../../lib/deno/config.ts";
import { ScenarioResult } from "../../lib/deno/runner.ts";
export { ScenarioResult, StepResult, StepStatus } from "../../lib/deno/runner.ts";
export type { ScenarioReport } from "../../lib/deno/runner.ts";
import { XrpcClient, XrpcError } from "../../lib/deno/client.ts";
import { assert } from "../../lib/deno/assertions.ts";
import { timedCall } from "../../lib/deno/runner.ts";

/**
 * Executes the scenario logic.
 * @returns A promise that resolves to the scenario result
 */

export async function run(): Promise<ScenarioResult> {
  const result = new ScenarioResult("Identity Fatigue");
  result.start();

  const client = new XrpcClient(PDS1);
  const rosa = getCharacter("rosa");

  await timedCall(result, "PDS health check", async () => {
    await client.waitForHealthy(30);
  });

  if (result.failed > 0) return result;

  const session = await timedCall(result, "Create account: rosa", async () => {
    return await client.accounts.createAccount(rosa.handle, rosa.email, rosa.password);
  });

  if (!session) {
    result.finish();
    return result;
  }
  rosa.did = session.did;
  rosa.accessJwt = session.accessJwt;

  const hourlyLimit = parseInt(Deno.env.get("PLC_HOURLY_LIMIT") || "5");
  const rotations = Math.min(hourlyLimit - 1, 10);

  let successCount = 0;
  for (let i = 0; i < rotations; i++) {
    try {
      const tokenResp = await client.raw.xrpcPost(
        "com.atproto.identity.requestPlcOperationSignature",
        {},
        rosa.accessJwt,
      );
      const signResp = await client.raw.xrpcPost("com.atproto.identity.signPlcOperation", {
        token: tokenResp.token,
        alsoKnownAs: [`at://rev-${i}-${rosa.handle}`],
      }, rosa.accessJwt);

      const op = { ...signResp.operation };
      delete op.did;

      const plcRes = await fetch(`${SERVICE_URLS.plc}/${rosa.did}`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(op),
      });

      if (plcRes.status === 200) {
        successCount++;
      } else {
        result.stepFailed("Exhaust Quota", `Failed at iteration ${i}: ${plcRes.status}`);
        break;
      }
    } catch (e) {
      result.stepFailed("Exhaust Quota", String(e));
      break;
    }
  }

  if (successCount === rotations) {
    result.stepPassed("Quota Exhaustion", `Successfully performed ${successCount} rotations`);

    // Final rotation should fail
    const tokenResp = await client.raw.xrpcPost(
      "com.atproto.identity.requestPlcOperationSignature",
      {},
      rosa.accessJwt,
    );
    const signResp = await client.raw.xrpcPost("com.atproto.identity.signPlcOperation", {
      token: tokenResp.token,
      alsoKnownAs: [`at://final-${rosa.handle}`],
    }, rosa.accessJwt);

    const op = { ...signResp.operation };
    delete op.did;

    const plcRes = await fetch(`${SERVICE_URLS.plc}/${rosa.did}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(op),
    });

    const body = await plcRes.text();
    if (plcRes.status === 400 && body.includes("Too many operations")) {
      result.stepPassed("Verify Hourly Limit", "Rejected operation after limit reached");
    } else {
      result.stepFailed(
        "Verify Hourly Limit",
        `Expected 400 rejection, got ${plcRes.status}: ${body}`,
      );
    }
  }

  result.finish();
  return result;
}

if (import.meta.main) {
  run().then((res) => {
    console.log(res.summary());
    Deno.exit(res.ok ? 0 : 1);
  });
}
